File: pkg/render/typst/typst.go
// Пакет typst малює diagram.Diagram у нативний Typst (через CeTZ). На відміну
// від SVG, результат — це Typst-код, що компілюється у векторну графіку без
// зовнішніх картинок. Координати точні (ті самі, що в SVG), тож вигляд збігається.
package typst

import (
	"fmt"
	"strings"

	"github.com/OlexiyOdarchuk/rombik/pkg/diagram"
)

// Render повертає САМОДОСТАТНІЙ Typst-документ: сторінка авто-розміру (щільно
// огортає схему — не A4 з обрізанням), за потреби — підпис «Рисунок N» через
// figure. Береш цей код і одразу компілюєш у тісний PDF.
func Render(d *diagram.Diagram) string {
	var b strings.Builder
	preamble(&b, d)
	figure(&b, d)
	return b.String()
}

// RenderAll повертає ОДИН Typst-документ з усіма схемами (кожна — окрема
// сторінка авто-розміру, figure з авто-нумерацією «Рисунок 1, 2, …»). Слово й
// роздільник підпису беремо з першої схеми (вони глобальні).
func RenderAll(ds []*diagram.Diagram) string {
	if len(ds) == 0 {
		return ""
	}
	var b strings.Builder
	preamble(&b, ds[0])
	for i, d := range ds {
		if i > 0 {
			b.WriteString("#pagebreak()\n")
		}
		figure(&b, d)
	}
	return b.String()
}

// preamble — шапка документа: import, авто-сторінка, шрифт, помічник #flowchart.
func preamble(b *strings.Builder, d *diagram.Diagram) {
	b.WriteString(`#import "@preview/cetz:0.3.4"` + "\n")
	b.WriteString("#set page(width: auto, height: auto, margin: 14pt)\n")
	b.WriteString("#set text(" + font + ")\n")
	fmt.Fprintf(b, "#set figure.caption(separator: [%s])\n", d.CapSeparator()) // напр. « — »
	supplement := "[" + d.CapSupplement() + "]"
	if !d.CapHasWord() { // шаблон без {word} — без слова-supplement
		supplement = "none"
	}
	b.WriteString(`#let flowchart(body, caption: none) = figure(` + "\n" +
		"  body, caption: caption, supplement: " + supplement + `, kind: "flowchart", numbering: "1",` + "\n)\n")
}

// figure — одна схема: cetz.canvas, за потреби загорнутий у #flowchart-підпис.
func figure(b *strings.Builder, d *diagram.Diagram) {
	canvas := renderCanvas(d)
	if d.Caption != "" {
		// #%q — рядкова форма, щоб спецсимволи Typst у підписі не ламали документ.
		fmt.Fprintf(b, "#flowchart(caption: [#%q])[\n%s]\n", d.Caption, canvas)
	} else {
		b.WriteString(canvas)
	}
}

// Fragment — лише блок cetz.canvas (без преамбули/сторінки/підпису), щоб
// вставити у свій .typ-документ (де import cetz уже є). Зручно для звіту.
func Fragment(d *diagram.Diagram) string { return renderCanvas(d) }

// FragmentAll — кілька canvas-фрагментів поспіль (через порожній рядок).
func FragmentAll(ds []*diagram.Diagram) string {
	var b strings.Builder
	for i, d := range ds {
		if i > 0 {
			b.WriteString("\n")
		}
		b.WriteString(renderCanvas(d))
	}
	return b.String()
}

// renderCanvas — сам блок cetz.canvas зі схемою (без сторінки/підпису).
func renderCanvas(d *diagram.Diagram) string {
	var b strings.Builder
	b.WriteString("#cetz.canvas(length: 1pt, {\n")
	b.WriteString("  import cetz.draw: *\n")
	b.WriteString("  set-style(stroke: 1.5pt, fill: none)\n")

	fy := func(y float64) float64 { return d.H - y } // CeTZ: вісь y вгору

	for _, e := range d.Edges {
		pts := make([]string, len(e.Points))
		for i, p := range e.Points {
			pts[i] = fmt.Sprintf("(%.1f, %.1f)", p.X, fy(p.Y))
		}
		mark := ""
		if !e.Arrowless {
			mark = `, mark: (end: ">")`
		}
		fmt.Fprintf(&b, "  line(%s%s)\n", strings.Join(pts, ", "), mark)
		if e.Label != "" && len(e.Points) >= 2 {
			lx, ly, align := diagram.LabelAnchor(e.Points[0], e.Points[1])
			anchor := "center"
			switch align {
			case "start":
				anchor = "west"
			case "end":
				anchor = "east"
			}
			fmt.Fprintf(&b, "  content((%.1f, %.1f), text(%ssize: 12pt)[#%q], anchor: %q)\n", lx, fy(ly), font, e.Label, anchor)
		}
	}
	for _, s := range d.Shapes {
		shape(&b, s, fy)
	}

	b.WriteString("})\n")
	return b.String()
}

func shape(b *strings.Builder, s diagram.Shape, fy func(float64) float64) {
	x1, y1, x2, y2 := s.X, s.Y, s.X+s.W, s.Y+s.H
	cx, cy := s.X+s.W/2, s.Y+s.H/2
	const fill = ", fill: white"
	poly := func(pts ...[2]float64) {
		parts := make([]string, len(pts))
		for i, p := range pts {
			parts[i] = fmt.Sprintf("(%.1f, %.1f)", p[0], p[1])
		}
		fmt.Fprintf(b, "  line(%s, close: true%s)\n", strings.Join(parts, ", "), fill)
	}

	switch s.Kind {
	case diagram.Terminator:
		fmt.Fprintf(b, "  rect((%.1f, %.1f), (%.1f, %.1f), radius: %.1f%s)\n", x1, fy(y2), x2, fy(y1), s.H/2, fill)
	case diagram.Process:
		fmt.Fprintf(b, "  rect((%.1f, %.1f), (%.1f, %.1f)%s)\n", x1, fy(y2), x2, fy(y1), fill)
	case diagram.Decision:
		poly([2]float64{cx, fy(y1)}, [2]float64{x2, fy(cy)}, [2]float64{cx, fy(y2)}, [2]float64{x1, fy(cy)})
	case diagram.InOut:
		sk := s.H * 0.4
		poly([2]float64{x1 + sk, fy(y1)}, [2]float64{x2, fy(y1)}, [2]float64{x2 - sk, fy(y2)}, [2]float64{x1, fy(y2)})
	case diagram.Hexagon:
		sk := s.H * 0.5
		poly([2]float64{x1 + sk, fy(y1)}, [2]float64{x2 - sk, fy(y1)}, [2]float64{x2, fy(cy)},
			[2]float64{x2 - sk, fy(y2)}, [2]float64{x1 + sk, fy(y2)}, [2]float64{x1, fy(cy)})
	case diagram.Predef:
		fmt.Fprintf(b, "  rect((%.1f, %.1f), (%.1f, %.1f)%s)\n", x1, fy(y2), x2, fy(y1), fill)
		fmt.Fprintf(b, "  line((%.1f, %.1f), (%.1f, %.1f))\n", x1+9, fy(y1), x1+9, fy(y2))
		fmt.Fprintf(b, "  line((%.1f, %.1f), (%.1f, %.1f))\n", x2-9, fy(y1), x2-9, fy(y2))
	case diagram.Connector:
		r := s.W / 2
		if s.H < s.W {
			r = s.H / 2
		}
		fmt.Fprintf(b, "  circle((%.1f, %.1f), radius: %.1f%s)\n", cx, fy(cy), r, fill)
	}
	// Текст по центру (рядкова форма #%q — щоб спецсимволи [ ] * не ламали Typst).
	fmt.Fprintf(b, "  content((%.1f, %.1f), text(%ssize: 14pt)[#%q])\n", cx, fy(cy), font, s.Text)
}

// font — Times New Roman (вимога курсових), із serif-фолбеками; перший наявний виграє.
const font = `font: ("Times New Roman", "Liberation Serif", "DejaVu Serif"), `

File: packages/engine/src/render/typst.ts
// Typst-рендерер (через CeTZ) — порт pkg/render/typst. Вивід — Typst-код, що
// компілюється у вектор. Координати ті самі, що в SVG (вісь y CeTZ — вгору, тож fy).
// Має збігатися БАЙТ-У-БАЙТ із Go (golden).
import type { Diagram, Shape } from '../diagram.ts';
import { labelAnchor, captionLine, capSupplement, capSeparator, capHasWord } from '../diagram.ts';
import { f1 as f } from '../format.ts';

const font = 'font: ("Times New Roman", "Liberation Serif", "DejaVu Serif"), ';

// goQuote — дзеркало Go strconv.Quote / fmt %q: "..." з екрануванням. Друковані
// Unicode-символи (кирилиця, «», оператори) лишаються як є.
function goQuote(s: string): string {
  let out = '"';
  for (const ch of s) {
    const c = ch.codePointAt(0) as number;
    switch (ch) {
      case '"': out += '\\"'; continue;
      case '\\': out += '\\\\'; continue;
      case '\n': out += '\\n'; continue;
      case '\r': out += '\\r'; continue;
      case '\t': out += '\\t'; continue;
      case '\x07': out += '\\a'; continue;
      case '\b': out += '\\b'; continue;
      case '\f': out += '\\f'; continue;
      case '\v': out += '\\v'; continue;
    }
    if (c < 0x20 || c === 0x7f) out += '\\x' + c.toString(16).padStart(2, '0');
    else out += ch;
  }
  return out + '"';
}

export function render(d: Diagram): string {
  return preamble(d) + figure(d);
}

export function renderAll(ds: Diagram[]): string {
  if (ds.length === 0) return '';
  let s = preamble(ds[0]);
  ds.forEach((d, i) => { if (i > 0) s += '#pagebreak()\n'; s += figure(d); });
  return s;
}

export function fragment(d: Diagram): string { return renderCanvas(d); }
export function fragmentAll(ds: Diagram[]): string {
  return ds.map(renderCanvas).join('\n');
}

function preamble(d: Diagram): string {
  let supplement = '[' + capSupplement(d) + ']';
  if (!capHasWord(d)) supplement = 'none';
  return '#import "@preview/cetz:0.3.4"\n' +
    '#set page(width: auto, height: auto, margin: 14pt)\n' +
    '#set text(' + font + ')\n' +
    `#set figure.caption(separator: [${capSeparator(d)}])\n` +
    '#let flowchart(body, caption: none) = figure(\n' +
    '  body, caption: caption, supplement: ' + supplement + ', kind: "flowchart", numbering: "1",\n)\n';
}

function figure(d: Diagram): string {
  const canvas = renderCanvas(d);
  if (d.caption) return `#flowchart(caption: [#${goQuote(d.caption)}])[\n${canvas}]\n`;
  return canvas;
}

function renderCanvas(d: Diagram): string {
  const fy = (y: number) => d.h - y;
  let b = '#cetz.canvas(length: 1pt, {\n  import cetz.draw: *\n  set-style(stroke: 1.5pt, fill: none)\n';
  for (const e of d.edges) {
    const pts = e.points.map((p) => `(${f(p.x)}, ${f(fy(p.y))})`).join(', ');
    const mark = e.arrowless ? '' : ', mark: (end: ">")';
    b += `  line(${pts}${mark})\n`;
    if (e.label && e.points.length >= 2) {
      const { x: lx, y: ly, align } = labelAnchor(e.points[0], e.points[1]);
      const anchor = align === 'start' ? 'west' : align === 'end' ? 'east' : 'center';
      b += `  content((${f(lx)}, ${f(fy(ly))}), text(${font}size: 12pt)[#${goQuote(e.label)}], anchor: ${goQuote(anchor)})\n`;
    }
  }
  for (const s of d.shapes) b += shape(s, fy);
  return b + '})\n';
}

function shape(s: Shape, fy: (y: number) => number): string {
  const x1 = s.x, y1 = s.y, x2 = s.x + s.w, y2 = s.y + s.h;
  const cx = s.x + s.w / 2, cy = s.y + s.h / 2;
  const fill = ', fill: white';
  const poly = (...pts: [number, number][]) =>
    `  line(${pts.map((p) => `(${f(p[0])}, ${f(p[1])})`).join(', ')}, close: true${fill})\n`;
  let b = '';
  switch (s.kind) {
    case 'terminator':
      b += `  rect((${f(x1)}, ${f(fy(y2))}), (${f(x2)}, ${f(fy(y1))}), radius: ${f(s.h / 2)}${fill})\n`;
      break;
    case 'process':
      b += `  rect((${f(x1)}, ${f(fy(y2))}), (${f(x2)}, ${f(fy(y1))})${fill})\n`;
      break;
    case 'decision':
      b += poly([cx, fy(y1)], [x2, fy(cy)], [cx, fy(y2)], [x1, fy(cy)]);
      break;
    case 'io': {
      const sk = s.h * 0.4;
      b += poly([x1 + sk, fy(y1)], [x2, fy(y1)], [x2 - sk, fy(y2)], [x1, fy(y2)]);
      break;
    }
    case 'loop': {
      const sk = s.h * 0.5;
      b += poly([x1 + sk, fy(y1)], [x2 - sk, fy(y1)], [x2, fy(cy)], [x2 - sk, fy(y2)], [x1 + sk, fy(y2)], [x1, fy(cy)]);
      break;
    }
    case 'subprogram':
      b += `  rect((${f(x1)}, ${f(fy(y2))}), (${f(x2)}, ${f(fy(y1))})${fill})\n`;
      b += `  line((${f(x1 + 9)}, ${f(fy(y1))}), (${f(x1 + 9)}, ${f(fy(y2))}))\n`;
      b += `  line((${f(x2 - 9)}, ${f(fy(y1))}), (${f(x2 - 9)}, ${f(fy(y2))}))\n`;
      break;
    case 'connector': {
      const r = s.h < s.w ? s.h / 2 : s.w / 2;
      b += `  circle((${f(cx)}, ${f(fy(cy))}), radius: ${f(r)}${fill})\n`;
      break;
    }
  }
  b += `  content((${f(cx)}, ${f(fy(cy))}), text(${font}size: 14pt)[#${goQuote(s.text)}])\n`;
  return b;
}
